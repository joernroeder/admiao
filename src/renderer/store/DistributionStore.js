import React, { createContext, useContext, useReducer, useEffect } from 'react'

const localStorageKey = 'DistributionState'

const initialState = {
	distributionType: undefined,
	distributionIdentifier: undefined,
}

const DistributionTypes = {
	LOCAL_PRINT: 'LOCAL_PRINT',
	REMOTE_PRINT: 'REMOTE_PRINT',
	DIGITAL: 'DIGITAL',
	FILE_SYSTEM: 'FILE_SYSTEM',
}

function reducer(state, action) {
	// sets Distribution Type
	if (Object.keys(DistributionTypes).includes(action.type)) {
		return {
			...state,
			distributionType: DistributionTypes[action.type],
			distributionIdentifier: action.payload || undefined,
		}
	}

	switch (action.type) {
		case 'RESET': {
			localStorage.removeItem(localStorageKey)

			return {
				...initialState,
			}
		}

		case 'SET_IDENTIFIER': {
			return {
				...state,
				distributionIdentifier: action.payload,
			}
		}

		default:
			return state
	}
}

const DistributionStateContext = createContext(undefined)
const DistributionDispatchContext = createContext(undefined)

function DistributionProvider({ children }) {
	const [state, dispatch] = useReducer(
		reducer,
		initialState,
		initialState => {
			const previousState = localStorage.getItem(localStorageKey)

			return previousState ? JSON.parse(previousState) : initialState
		}
	)

	// store distributionType in local storage.
	// This stores the distribution type to improve the backup process.
	useEffect(() => {
		if (state.isInitialState) {
			return
		}

		// filter out empty and initial states
		if (!Object.values(state).filter(Boolean).length) {
			return
		}

		console.log(Object.values(state).filter(Boolean).length)
		localStorage.setItem(localStorageKey, JSON.stringify(state))
	}, [state])

	return (
		<DistributionStateContext.Provider value={state}>
			<DistributionDispatchContext.Provider value={dispatch}>
				{children}
			</DistributionDispatchContext.Provider>
		</DistributionStateContext.Provider>
	)
}

function useDistributionState() {
	const context = useContext(DistributionStateContext)

	if (context === undefined) {
		throw new Error(
			'useDistributionState must be used within a DistributionProvider'
		)
	}
	return context
}

function useDistributionDispatch() {
	const context = useContext(DistributionDispatchContext)

	if (context === undefined) {
		throw new Error(
			'useDistributionDispatch must be used within a DistributionProvider'
		)
	}
	return context
}

export {
	DistributionProvider,
	useDistributionState,
	useDistributionDispatch,
	DistributionTypes,
}
