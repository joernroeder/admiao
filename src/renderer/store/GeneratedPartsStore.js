import React, { createContext, useContext, useReducer } from 'react'

const initialState = {
	parts: [],
	confirmed: [],
}

function reducer(state, action) {
	switch (action.type) {
		case 'SET_PARTS': {
			console.log('setting parts in store', action)
			return {
				...state,
				confirmed: Array(action.payload.length).fill(false),
				parts: action.payload,
			}
		}

		case 'CONFIRMED': {
			const updatedConfirmed = {
				confirmed: state.confirmed.map((isConfirmed, index) => {
					if (index === action.payload) {
						return true
					}

					return isConfirmed
				}),
			}

			return {
				...state,
				...updatedConfirmed,
			}
		}

		case 'RESET': {
			return {
				...initialState,
			}
		}

		default:
			return state
	}
}

const GeneratedPartsStateContext = createContext(undefined)
const GeneratedPartsDispatchContext = createContext(undefined)

function GeneratedPartsProvider({ children }) {
	const [state, dispatch] = useReducer(reducer, initialState)

	return (
		<GeneratedPartsStateContext.Provider value={state}>
			<GeneratedPartsDispatchContext.Provider value={dispatch}>
				{children}
			</GeneratedPartsDispatchContext.Provider>
		</GeneratedPartsStateContext.Provider>
	)
}

function useGeneratedParts() {
	const context = useContext(GeneratedPartsStateContext)

	if (context === undefined) {
		throw new Error(
			'useGeneratedParts must be used within a GeneratedPartsProvider'
		)
	}
	return context
}

function useGeneratedPartsDispatch() {
	const context = useContext(GeneratedPartsDispatchContext)

	if (context === undefined) {
		throw new Error(
			'useDistributionDispatch must be used within a GeneratedPartsProvider'
		)
	}
	return context
}

export { GeneratedPartsProvider, useGeneratedParts, useGeneratedPartsDispatch }
