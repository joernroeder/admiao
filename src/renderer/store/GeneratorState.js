import React, { createContext, useContext, useState } from 'react'

const STATES = {
	loading: 'LOADING',
	successfullyGenerated: 'SUCCESSFULLY_GENERATED',
	generationFailed: 'GENERATION_FAILED',
	successfullyDistributed: 'SUCCESSFULLY_DISTRIBUTED',
	distributionFailed: 'DISTRIBUTION_FAILED',
}

const GeneratorStateContext = createContext(undefined)

function GeneratorStateProvider({ children }) {
	const [generatorState, setGeneratorState] = useState({
		state: undefined,
		pendingIdentifier: null,
		error: null,
		// additional data returned by the ipc methods may get stored here.
		// todo be more precise/refactor once all distribution types are implemented
		meta: {},
	})

	return (
		<GeneratorStateContext.Provider
			value={[generatorState, setGeneratorState]}
		>
			{children}
		</GeneratorStateContext.Provider>
	)
}

function useGeneratorState() {
	const context = useContext(GeneratorStateContext)

	if (context === undefined) {
		throw new Error(
			'useGeneratedParts must be used within a GeneratedPartsProvider'
		)
	}
	return context
}

export { GeneratorStateProvider, useGeneratorState, STATES }
