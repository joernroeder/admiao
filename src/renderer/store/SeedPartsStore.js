import React, { createContext, useContext, useEffect, useReducer } from 'react'

const localStorageKey = 'SeedPartsState'

const initialState = {
	uniquePartsN: 3,
	requiredPartsT: 2,

	partsPerDisk: 1,
	requiredDisks: 3,

	recoveredFromPreviousSelection: false,
	isInitialState: true,
}

const calcDisks = (uniquePartsN, requiredPartsT) => {
	const partsPerDisk = requiredPartsT - 1
	const requiredDisks = Math.ceil(uniquePartsN / partsPerDisk)

	return {
		partsPerDisk,
		requiredDisks,
	}
}

function reducer(state, action) {
	const { uniquePartsN, requiredPartsT } = state

	const isModifiedState = {
		recoveredFromPreviousSelection: false,
		isInitialState: false,
	}

	switch (action.type) {
		case 'INCREMENT_REQUIRED': {
			const updatedRequiredPartsT = Math.min(
				requiredPartsT + 1,
				uniquePartsN
			)

			return {
				...state,
				requiredPartsT: updatedRequiredPartsT,
				...calcDisks(state.uniquePartsN, updatedRequiredPartsT),
				...isModifiedState,
			}
		}

		case 'DECREMENT_REQUIRED': {
			const updatedRequiredPartsT = Math.max(requiredPartsT - 1, 1)

			return {
				...state,
				requiredPartsT: updatedRequiredPartsT,
				...calcDisks(state.uniquePartsN, updatedRequiredPartsT),
				...isModifiedState,
			}
		}

		case 'INCREMENT_UNIQUE': {
			const updatedUniquePartsN = uniquePartsN + 1

			return {
				...state,
				uniquePartsN: updatedUniquePartsN,
				...calcDisks(updatedUniquePartsN, state.requiredPartsT),
				...isModifiedState,
			}
		}

		case 'DECREMENT_UNIQUE': {
			const updatedUniquePartsN = Math.max(uniquePartsN - 1, 1)
			// also decrement required parts if they would exceed otherwise
			const updatedRequiredPartsT = Math.min(
				updatedUniquePartsN,
				requiredPartsT
			)

			return {
				...state,
				uniquePartsN: updatedUniquePartsN,
				requiredPartsT: updatedRequiredPartsT,
				...calcDisks(updatedUniquePartsN, updatedRequiredPartsT),
				...isModifiedState,
			}
		}

		case 'CONFIRM': {
			return {
				...state,
				...isModifiedState,
			}
		}

		case 'RESET': {
			localStorage.removeItem(localStorageKey)

			return {
				...initialState,
			}
		}

		default:
			return state
	}
}

const SeedPartsStateContext = createContext(undefined)
const SeedPartsDispatchContext = createContext(undefined)

function SeedPartsProvider({ children }) {
	// uses a previous state from local storage if available or uses the initial one defined above.
	const [state, dispatch] = useReducer(
		reducer,
		initialState,
		initialState => {
			const previousState = localStorage.getItem(localStorageKey)

			return previousState
				? {
						...JSON.parse(previousState),
						recoveredFromPreviousSelection: true,
				  }
				: initialState
		}
	)

	// store current state in local storage.
	// This is used to improve the backup process if the user paused to print the
	// seed templates.
	useEffect(() => {
		if (state.isInitialState) {
			return
		}

		localStorage.setItem(localStorageKey, JSON.stringify(state))
	}, [state])

	return (
		<SeedPartsStateContext.Provider value={state}>
			<SeedPartsDispatchContext.Provider value={dispatch}>
				{children}
			</SeedPartsDispatchContext.Provider>
		</SeedPartsStateContext.Provider>
	)
}

function useSeedPartsState() {
	const context = useContext(SeedPartsStateContext)

	if (context === undefined) {
		throw new Error(
			'useSeedPartsState must be used within a SeedPartsProvider'
		)
	}
	return context
}

function useSeedPartsDispatch() {
	const context = useContext(SeedPartsDispatchContext)

	if (context === undefined) {
		throw new Error(
			'useSeedPartsDispatch must be used within a SeedPartsProvider'
		)
	}
	return context
}

export { SeedPartsProvider, useSeedPartsState, useSeedPartsDispatch }
