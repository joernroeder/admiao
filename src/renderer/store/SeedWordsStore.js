import React, { createContext, useContext, useReducer } from 'react'
const bip39 = require('bip39')

const initialWords = [
	/*
	'answer',
	'dice',
	'warm',
	'cup',
	'shock',
	'orbit',
	'pistol',
	'seat',
	'modify',
	'matrix',
	'genre',
	'renew',
	*/
]

const initialState = {
	words: initialWords,
	total: 12, // todo handle other seeds 18, 24
	isConfirmed: false,
	isValid: getIsValid(initialWords),
}

function getIsValid(words) {
	return bip39.validateMnemonic(words.join(' '))
}

function wordsReducer(state, action) {
	const { words } = state

	switch (action.type) {
		case 'APPEND': {
			const updatedWords = [...words, action.payload]

			return {
				...state,
				words: updatedWords,
				isValid: getIsValid(updatedWords),
			}
		}

		case 'RESET': {
			return {
				...state,
				words: [],
				isConfirmed: false,
				isValid: false,
			}
		}

		case 'CLEANUP_SEED': {
			return {
				...state,
				words: [],
			}
		}

		case 'POP': {
			const updatedWords = [...words.slice(0, -1)]

			return {
				...state,
				words: updatedWords,
				isValid: getIsValid(updatedWords),
			}
		}

		case 'CONFIRM': {
			return {
				...state,
				isConfirmed: true,
			}
		}

		default:
			return state
	}
}

const SeedWordsStateContext = createContext(undefined)
const SeedWordsDispatchContext = createContext(undefined)

function SeedWordsProvider({ children }) {
	const [state, dispatch] = useReducer(wordsReducer, initialState)

	return (
		<SeedWordsStateContext.Provider value={state}>
			<SeedWordsDispatchContext.Provider value={dispatch}>
				{children}
			</SeedWordsDispatchContext.Provider>
		</SeedWordsStateContext.Provider>
	)
}

function useSeedWordsState() {
	const context = useContext(SeedWordsStateContext)

	if (context === undefined) {
		throw new Error(
			'useSeedWordsState must be used within a SeedWordsProvider'
		)
	}
	return context
}

function useSeedWordsDispatch() {
	const context = useContext(SeedWordsDispatchContext)

	if (context === undefined) {
		throw new Error(
			'useSeedWordsDispatch must be used within a SeedWordsProvider'
		)
	}
	return context
}

export {
	SeedWordsProvider,
	useSeedWordsState,
	useSeedWordsDispatch,
	wordsReducer,
}
