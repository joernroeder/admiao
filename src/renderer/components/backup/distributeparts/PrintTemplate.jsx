import React, { useState, useEffect } from 'react'
import { Flex } from 'reflexbox'

import { useDistributionDispatch } from '../../../store/DistributionStore'

import GridWrap from '../../GridWrap'
import Cell from '../../Cell'
import SubHeading from '../../SubHeading'
import Text from '../../Text'
import Button from '../../Button'

const PrintTemplate = ({
	templateIdentifier,
	onCopyToDrive,
	templateError,
}) => {
	const { ipcRenderer } = window
	const [templateOpened, setTemplateOpened] = useState(false)
	const distributionDispatch = useDistributionDispatch()

	const openTemplateInWindow = () => {
		ipcRenderer.send('open-template', { identifier: templateIdentifier })
	}

	useEffect(() => {
		if (!templateIdentifier) {
			return
		}

		const onTemplateOpened = (event, args) => {
			const { identifier, isOpen } = args

			if (identifier !== templateIdentifier) {
				return
			}

			if (isOpen) {
				setTemplateOpened(true)

				// push the identifier to the distribution store.
				// this is used later to match generated parts with a distribution selection.
				distributionDispatch({
					type: 'SET_IDENTIFIER',
					payload: templateIdentifier,
				})
			} else {
				console.warn('ERROR: could not open file', identifier)
			}
		}

		ipcRenderer.once('template-opened', onTemplateOpened)

		return () => {
			ipcRenderer.removeListener('template-opened', onTemplateOpened)
		}
	}, [distributionDispatch, ipcRenderer, templateIdentifier])

	return (
		<GridWrap flexDirection={'column'}>
			{templateOpened ? (
				<>
					<Cell gridOffset={4} gridColumn={4} mt={9}>
						<SubHeading mb={2}>Print Templates</SubHeading>
					</Cell>
					<Cell gridOffset={4} gridColumn={4} mt={3}>
						<Text>
							Please print out the PDF template and continue here
							afterwards.
						</Text>

						<Button
							mt={3}
							to={'../../enter-seed'}
							variant={'filled'}
						>
							Yes, I've printed the templates
						</Button>
					</Cell>
				</>
			) : (
				<>
					<Cell gridOffset={4} gridColumn={4} mt={9}>
						<SubHeading mb={2}>Print Templates</SubHeading>
					</Cell>
					<Cell gridOffset={4} gridColumn={4} mt={3}>
						<Text>Is your printer connected to this device?</Text>

						<Flex
							flexDirection={'column'}
							alignItems={'flex-end'}
							mt={3}
						>
							<Button onClick={onCopyToDrive} mb={1}>
								No, copy Templates-PDF to a USB-Drive
							</Button>
							<Button onClick={openTemplateInWindow}>
								Yes, open Templates-PDF
							</Button>
						</Flex>
					</Cell>
				</>
			)}
		</GridWrap>
	)
}

export default PrintTemplate
