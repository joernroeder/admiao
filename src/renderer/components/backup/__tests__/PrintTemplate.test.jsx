import React from 'react'

import { act, fireEvent, render } from '@testing-library/react'
import { DistributionProvider } from '../../../store/DistributionStore'

import PrintTemplate from '../distributeparts/PrintTemplate'

describe('Print Templates', () => {
	beforeEach(() => {
		global.ipcRenderer = {
			send: jest.fn(),
			once: jest.fn(),
			removeListener: jest.fn(),
		}
	})

	const renderWithProviders = (children, renderFunction = render) => {
		return renderFunction(
			<DistributionProvider>{children}</DistributionProvider>
		)
	}

	test('it should render correctly', () => {
		const { getByText, container } = renderWithProviders(<PrintTemplate />)

		expect(getByText('Print Templates')).toBeInTheDocument()

		expect(
			getByText(/No, copy Templates-PDF/i, { exact: false })
		).toBeInTheDocument()
		expect(
			getByText(/Yes, open Templates-PDF/i, { exact: false })
		).toBeInTheDocument()

		expect(container.firstChild).toMatchSnapshot()
	})

	describe('open template via external app', () => {
		test('it should send a `open-template` message via ipc on button click', () => {
			const { getByText, unmount } = renderWithProviders(
				<PrintTemplate templateIdentifier={'abc'} />
			)

			fireEvent.click(getByText(/Yes, open/i))

			expect(global.ipcRenderer.send).toBeCalledTimes(1)
			expect(global.ipcRenderer.send).toHaveBeenCalledWith(
				'open-template',
				{ identifier: 'abc' }
			)

			// register listener
			expect(global.ipcRenderer.once).toHaveBeenCalledWith(
				'template-opened',
				expect.any(Function)
			)

			// test cleanup
			unmount()
			expect(global.ipcRenderer.removeListener).toHaveBeenCalledWith(
				'template-opened',
				expect.any(Function)
			)
		})

		describe('it should update once it received `template-opened` message', () => {
			test('valid response', () => {
				const { getByText, container } = renderWithProviders(
					<PrintTemplate templateIdentifier={'abc'} />
				)
				const [
					channel,
					callback,
				] = global.ipcRenderer.once.mock.calls[0]

				expect(channel).toEqual('template-opened')

				act(() => {
					callback(null, { identifier: 'abc', isOpen: true })
				})

				expect(
					getByText(/printed the templates/i, { exact: false })
				).toBeInTheDocument()

				expect(container.firstChild).toMatchSnapshot()
			})

			test('invalid identifier', () => {
				const { getByText } = renderWithProviders(
					<PrintTemplate templateIdentifier={'abc'} />
				)
				const [_, callback] = global.ipcRenderer.once.mock.calls[0]

				act(() => {
					callback(null, { identifier: 'some-other', isOpen: true })
				})

				expect(
					getByText(/Yes, open Templates-PDF/i, { exact: false })
				).toBeInTheDocument()
			})

			// todo, return to the user that the pdf could not be opened by the external app
			test('could not open', () => {
				const { getByText } = renderWithProviders(
					<PrintTemplate templateIdentifier={'abc'} />
				)
				const [_, callback] = global.ipcRenderer.once.mock.calls[0]

				act(() => {
					callback(null, { identifier: 'abc', isOpen: false })
				})

				expect(
					getByText(/Yes, open Templates-PDF/i, { exact: false })
				).toBeInTheDocument()
			})
		})

		test('it should call the passed in Â´onCopyToDrive` function', () => {
			const onCopyToDrive = jest.fn()
			const { getByText } = renderWithProviders(
				<PrintTemplate onCopyToDrive={onCopyToDrive} />
			)

			fireEvent.click(getByText(/No, copy Templates/i, { exact: false }))

			expect(onCopyToDrive).toBeCalledTimes(1)
		})
	})
})
