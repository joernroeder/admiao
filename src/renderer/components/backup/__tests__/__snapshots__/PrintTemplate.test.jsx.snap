// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Print Templates it should render correctly 1`] = `
.emotion-9 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  margin-left: -4px;
  margin-right: -4px;
  background-color: white;
  color: black;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.emotion-1 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  margin-top: 9px;
  padding-left: 4px;
  padding-right: 4px;
  width: 33.33333333333333%;
}

.emotion-0 {
  color: black;
  margin-bottom: 8px;
  margin-top: 0;
  font-size: 14px;
  line-height: 1;
  -webkit-transform: 1px;
  -ms-transform: 1px;
  transform: 1px;
  font-family: 'SpaceGrotesk',sans-serif;
  -webkit-letter-spacing: 0.01em;
  -moz-letter-spacing: 0.01em;
  -ms-letter-spacing: 0.01em;
  letter-spacing: 0.01em;
  text-rendering: geometricPrecision;
  -webkit-font-smoothing: antialiased;
  font-variant-ligatures: common-ligatures;
}

@media screen and (min-width:40em) {
  .emotion-0 {
    font-size: 16px;
    line-height: 2;
    -webkit-transform: 2px;
    -ms-transform: 2px;
    transform: 2px;
  }
}

@media screen and (min-width:52em) {
  .emotion-0 {
    font-size: 16px;
    line-height: 2;
    -webkit-transform: 2px;
    -ms-transform: 2px;
    transform: 2px;
  }
}

.emotion-8 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  margin-top: 16px;
  padding-left: 4px;
  padding-right: 4px;
  width: 33.33333333333333%;
}

.emotion-2 {
  color: black;
  margin-top: 0;
  margin-bottom: 0;
  font-size: 12px;
  line-height: 0;
  font-size: 12px;
  line-height: 0;
  -webkit-transform: 0;
  -ms-transform: 0;
  transform: 0;
  text-rendering: optimizeLegibility;
  -webkit-font-smoothing: antialiased;
  font-variant-ligatures: common-ligatures;
}

.emotion-7 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: flex-end;
  -webkit-box-align: flex-end;
  -ms-flex-align: flex-end;
  align-items: flex-end;
  margin-top: 16px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.emotion-4 {
  border-width: 2px;
  border-color: black;
  border-style: solid;
  margin-bottom: 4px;
  padding-left: 8px;
  padding-right: 8px;
  padding-top: 4px;
  padding-bottom: 4px;
  font-size: 12px;
  line-height: 0;
  -webkit-letter-spacing: 0.025em;
  -moz-letter-spacing: 0.025em;
  -ms-letter-spacing: 0.025em;
  letter-spacing: 0.025em;
  font-size: 12px;
  line-height: 0;
  -webkit-transform: 0;
  -ms-transform: 0;
  transform: 0;
  color: black;
  background-color: transparent;
  display: inline-block;
  -webkit-text-decoration: none;
  text-decoration: none;
  font-weight: 700;
  box-sizing: border-box;
  position: relative;
  z-index: 10;
  overflow: hidden;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  text-rendering: geometricPrecision;
  -webkit-font-smoothing: antialiased;
  cursor: pointer;
}

.emotion-4 .icon {
  display: inline-block;
  -webkit-transition: -webkit-transform 0.25s ease-in;
  -webkit-transition: transform 0.25s ease-in;
  transition: transform 0.25s ease-in;
}

.emotion-4:hover .icon {
  -webkit-transition-timing-function: ease-out;
  transition-timing-function: ease-out;
  -webkit-transition-duration: 0.15s;
  transition-duration: 0.15s;
}

.emotion-4:hover .icon.next {
  -webkit-transform: translateX(0.4em);
  -ms-transform: translateX(0.4em);
  transform: translateX(0.4em);
}

.emotion-4:hover .icon.prev {
  -webkit-transform: translateX(-0.4em);
  -ms-transform: translateX(-0.4em);
  transform: translateX(-0.4em);
}

.emotion-4:hover .icon.flip {
  -webkit-transform: rotate(90deg);
  -ms-transform: rotate(90deg);
  transform: rotate(90deg);
}

.emotion-4 > span:first-of-type {
  position: relative;
  z-index: 10;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

.emotion-3 {
  background-color: blackTransparent;
  display: block;
  position: absolute;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  opacity: 0;
  -webkit-transform: skew( -14deg, 0 ) translateX( -110% ) translateY( 0% );
  -ms-transform: skew( -14deg, 0 ) translateX( -110% ) translateY( 0% );
  transform: skew( -14deg, 0 ) translateX( -110% ) translateY( 0% );
  -webkit-transition: -webkit-transform 0.25s ease-in,opacity 0.25s ease-in;
  -webkit-transition: transform 0.25s ease-in,opacity 0.25s ease-in;
  transition: transform 0.25s ease-in,opacity 0.25s ease-in;
}

*:hover > .emotion-3,
*:focus > .emotion-3 {
  opacity: 1;
  -webkit-transition-timing-function: ease-out;
  transition-timing-function: ease-out;
  -webkit-transition-duration: 0.15s;
  transition-duration: 0.15s;
  -webkit-transform: translateX( -1.9em ) translateY(0) skew( -14deg, 0 );
  -ms-transform: translateX( -1.9em ) translateY(0) skew( -14deg, 0 );
  transform: translateX( -1.9em ) translateY(0) skew( -14deg, 0 );
}

.emotion-6 {
  border-width: 2px;
  border-color: black;
  border-style: solid;
  padding-left: 8px;
  padding-right: 8px;
  padding-top: 4px;
  padding-bottom: 4px;
  font-size: 12px;
  line-height: 0;
  -webkit-letter-spacing: 0.025em;
  -moz-letter-spacing: 0.025em;
  -ms-letter-spacing: 0.025em;
  letter-spacing: 0.025em;
  font-size: 12px;
  line-height: 0;
  -webkit-transform: 0;
  -ms-transform: 0;
  transform: 0;
  color: black;
  background-color: transparent;
  display: inline-block;
  -webkit-text-decoration: none;
  text-decoration: none;
  font-weight: 700;
  box-sizing: border-box;
  position: relative;
  z-index: 10;
  overflow: hidden;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  text-rendering: geometricPrecision;
  -webkit-font-smoothing: antialiased;
  cursor: pointer;
}

.emotion-6 .icon {
  display: inline-block;
  -webkit-transition: -webkit-transform 0.25s ease-in;
  -webkit-transition: transform 0.25s ease-in;
  transition: transform 0.25s ease-in;
}

.emotion-6:hover .icon {
  -webkit-transition-timing-function: ease-out;
  transition-timing-function: ease-out;
  -webkit-transition-duration: 0.15s;
  transition-duration: 0.15s;
}

.emotion-6:hover .icon.next {
  -webkit-transform: translateX(0.4em);
  -ms-transform: translateX(0.4em);
  transform: translateX(0.4em);
}

.emotion-6:hover .icon.prev {
  -webkit-transform: translateX(-0.4em);
  -ms-transform: translateX(-0.4em);
  transform: translateX(-0.4em);
}

.emotion-6:hover .icon.flip {
  -webkit-transform: rotate(90deg);
  -ms-transform: rotate(90deg);
  transform: rotate(90deg);
}

.emotion-6 > span:first-of-type {
  position: relative;
  z-index: 10;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

<div
  class="emotion-9"
>
  <div
    class="emotion-1"
  >
    <h2
      class="emotion-0"
      color="black"
      font-size="1,2,2"
    >
      Print Templates
    </h2>
  </div>
  <div
    class="emotion-8"
  >
    <p
      class="emotion-2"
      color="black"
      font-size="0"
    >
      Is your printer connected to this device?
    </p>
    <div
      class="emotion-7"
    >
      <button
        class="emotion-4"
      >
        <span>
          No, copy Templates-PDF to a USB-Drive
           
          <span
            class="icon next"
          >
            →
          </span>
        </span>
        <span
          class="emotion-3"
        />
      </button>
      <button
        class="emotion-6"
      >
        <span>
          Yes, open Templates-PDF
           
          <span
            class="icon next"
          >
            →
          </span>
        </span>
        <span
          class="emotion-3"
        />
      </button>
    </div>
  </div>
</div>
`;

exports[`Print Templates open template via external app it should update once it received \`template-opened\` message valid response 1`] = `
.emotion-6 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  margin-left: -4px;
  margin-right: -4px;
  background-color: white;
  color: black;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.emotion-1 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  margin-top: 9px;
  padding-left: 4px;
  padding-right: 4px;
  width: 33.33333333333333%;
}

.emotion-0 {
  color: black;
  margin-bottom: 8px;
  margin-top: 0;
  font-size: 14px;
  line-height: 1;
  -webkit-transform: 1px;
  -ms-transform: 1px;
  transform: 1px;
  font-family: 'SpaceGrotesk',sans-serif;
  -webkit-letter-spacing: 0.01em;
  -moz-letter-spacing: 0.01em;
  -ms-letter-spacing: 0.01em;
  letter-spacing: 0.01em;
  text-rendering: geometricPrecision;
  -webkit-font-smoothing: antialiased;
  font-variant-ligatures: common-ligatures;
}

@media screen and (min-width:40em) {
  .emotion-0 {
    font-size: 16px;
    line-height: 2;
    -webkit-transform: 2px;
    -ms-transform: 2px;
    transform: 2px;
  }
}

@media screen and (min-width:52em) {
  .emotion-0 {
    font-size: 16px;
    line-height: 2;
    -webkit-transform: 2px;
    -ms-transform: 2px;
    transform: 2px;
  }
}

.emotion-5 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  margin-top: 16px;
  padding-left: 4px;
  padding-right: 4px;
  width: 33.33333333333333%;
}

.emotion-2 {
  color: black;
  margin-top: 0;
  margin-bottom: 0;
  font-size: 12px;
  line-height: 0;
  font-size: 12px;
  line-height: 0;
  -webkit-transform: 0;
  -ms-transform: 0;
  transform: 0;
  text-rendering: optimizeLegibility;
  -webkit-font-smoothing: antialiased;
  font-variant-ligatures: common-ligatures;
}

.emotion-4 {
  border-width: 2px;
  border-color: black;
  border-style: solid;
  margin-top: 16px;
  padding-left: 8px;
  padding-right: 8px;
  padding-top: 4px;
  padding-bottom: 4px;
  font-size: 12px;
  line-height: 0;
  -webkit-letter-spacing: 0.025em;
  -moz-letter-spacing: 0.025em;
  -ms-letter-spacing: 0.025em;
  letter-spacing: 0.025em;
  font-size: 12px;
  line-height: 0;
  -webkit-transform: 0;
  -ms-transform: 0;
  transform: 0;
  color: white;
  background-color: black;
  display: inline-block;
  -webkit-text-decoration: none;
  text-decoration: none;
  font-weight: 700;
  box-sizing: border-box;
  position: relative;
  z-index: 10;
  overflow: hidden;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  text-rendering: geometricPrecision;
  -webkit-font-smoothing: antialiased;
  cursor: pointer;
}

.emotion-4 .icon {
  display: inline-block;
  -webkit-transition: -webkit-transform 0.25s ease-in;
  -webkit-transition: transform 0.25s ease-in;
  transition: transform 0.25s ease-in;
}

.emotion-4:hover .icon {
  -webkit-transition-timing-function: ease-out;
  transition-timing-function: ease-out;
  -webkit-transition-duration: 0.15s;
  transition-duration: 0.15s;
}

.emotion-4:hover .icon.next {
  -webkit-transform: translateX(0.4em);
  -ms-transform: translateX(0.4em);
  transform: translateX(0.4em);
}

.emotion-4:hover .icon.prev {
  -webkit-transform: translateX(-0.4em);
  -ms-transform: translateX(-0.4em);
  transform: translateX(-0.4em);
}

.emotion-4:hover .icon.flip {
  -webkit-transform: rotate(90deg);
  -ms-transform: rotate(90deg);
  transform: rotate(90deg);
}

.emotion-4 > span:first-of-type {
  position: relative;
  z-index: 10;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

.emotion-3 {
  background-color: whiteTransparent;
  display: block;
  position: absolute;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  opacity: 0;
  -webkit-transform: skew( -14deg, 0 ) translateX( -110% ) translateY( 0% );
  -ms-transform: skew( -14deg, 0 ) translateX( -110% ) translateY( 0% );
  transform: skew( -14deg, 0 ) translateX( -110% ) translateY( 0% );
  -webkit-transition: -webkit-transform 0.25s ease-in,opacity 0.25s ease-in;
  -webkit-transition: transform 0.25s ease-in,opacity 0.25s ease-in;
  transition: transform 0.25s ease-in,opacity 0.25s ease-in;
}

*:hover > .emotion-3,
*:focus > .emotion-3 {
  opacity: 1;
  -webkit-transition-timing-function: ease-out;
  transition-timing-function: ease-out;
  -webkit-transition-duration: 0.15s;
  transition-duration: 0.15s;
  -webkit-transform: translateX( -1.9em ) translateY(0) skew( -14deg, 0 );
  -ms-transform: translateX( -1.9em ) translateY(0) skew( -14deg, 0 );
  transform: translateX( -1.9em ) translateY(0) skew( -14deg, 0 );
}

<div
  class="emotion-6"
>
  <div
    class="emotion-1"
  >
    <h2
      class="emotion-0"
      color="black"
      font-size="1,2,2"
    >
      Print Templates
    </h2>
  </div>
  <div
    class="emotion-5"
  >
    <p
      class="emotion-2"
      color="black"
      font-size="0"
    >
      Please print out the PDF template and continue here afterwards.
    </p>
    <a
      class="emotion-4"
      href="/enter-seed"
    >
      <span>
        Yes, I've printed the templates
         
        <span
          class="icon next"
        >
          →
        </span>
      </span>
      <span
        class="emotion-3"
      />
    </a>
  </div>
</div>
`;
